src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginHandler.cs

```diff
@@ -0,0 +1,150 @@
+
+            InitializeSslStream(context);
+
+            IHandler<PreloginHandlerContext> firstHandler = _subHandlerChainBuilder.CreateChain(context, _tlsAuthenticator);
+            
+            await firstHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(connectionContext, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Initializes the SSL required for the TLS handshake.
+        /// In case of Tds7.4, the SslOverTdsStream is created as well.
+        /// </summary>
+        /// <param name="preloginContext"></param>
+        void InitializeSslStream(PreloginHandlerContext preloginContext)
+        {
+            // Create the streams
+            // If tls first then create a sslStream with the underlying stream as the transport stream.
+            // if this is not tlsfirst then ssl over tds stream with transport stream as the underlying stream.
+            Stream transportStream = preloginContext.ConnectionContext.ConnectionStream;
+
+            // baseStream is the underlying stream for the SslStream.
+            Stream baseStream = transportStream;
+            if (!preloginContext.IsTlsFirst)
+            {
+                SslOverTdsStream sslOVerTdsStream;
+                baseStream = sslOVerTdsStream 
+                    = preloginContext.ConnectionContext.SslOverTdsStream
+                    = new SslOverTdsStream(transportStream, preloginContext.ConnectionContext.ConnectionId);
+            }
+
+            SslStream sslStream =
+                preloginContext.ConnectionContext.SslStream 
+                = new SslStream(baseStream, true, ValidateServerCertificate);
+
+            Stream preloginStream = preloginContext.IsTlsFirst ? (Stream)sslStream : (Stream)preloginContext.ConnectionContext.ConnectionStream;
+
+            preloginContext.ConnectionContext.TdsStream = new IO.TdsStream(new IO.TdsWriteStream(preloginStream), new IO.TdsReadStream(preloginStream));
+
+            bool ValidateServerCertificate(object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
+            {
+                Guid connectionId = preloginContext.ConnectionContext.ConnectionId;
+                if (!preloginContext.ShouldValidateCertificate())
+                {
+                    SqlClientEventSource.Log.TrySNITraceEvent(nameof(PreloginHandler), EventType.INFO, "Connection Id {0}, Certificate will not be validated.", args0: connectionId);
+                    return true;
+                }
+
+                SqlClientEventSource.Log.TrySNITraceEvent(nameof(PreloginHandler), EventType.INFO, "Connection Id {0}, Certificate will be validated for Target Server name", args0: connectionId);
+
+                return SNICommon.ValidateSslServerCertificate(connectionId,
```

Comment:
Not critical at the moment, but these dependencies on Managed SNI can make things complicated for both SqlClient and SqlClientX, so I would recommend adding a comment in docs in all such APIs in Managed SNI, that they are shared APIs, and should be handled carefully.